[flang-rt] Add the ability to have user supplied callback functions to further customize the runtime environment.

d-parks opened the pull request:
Add the ability to have pre and post call back functions to ExecutionEnvironment::Configure() to allow further customization of the flang runtime environment (called from _FortranAStartProgam) in situations where either the desired features/functionality are proprietary or are too specific to be accepted by the flang community.

Example:
Custom constructor object linked with flang objects:
```
#include "flang-rt/runtime/environment.h"
#include "flang/Runtime/entry-names.h"
#include "flang/Runtime/extensions.h"

namespace Fortran::runtime {

// Do something specific to the flang runtime environment prior to the
// core logic of ExecutionEnvironment::Configure().
static void
CustomPreConfigureEnv(int argc, const char *argv[], const char *envp[],
                      const EnvironmentDefaultList *envDefaultList) {
  puts(__func__);
}

// Do something specific to the flang runtime environment after running the
// core logic of ExecutionEnvironment::Configure().
static void
CustomPostConfigureEnv(int argc, const char *argv[], const char *envp[],
                       const EnvironmentDefaultList *envDefaultList) {
  puts(__func__);
}

void __attribute__((constructor)) CustomInitCstor(void) {
  // Possibilities:
  // RTNAME(RegisterConfigureEnv)(&CustomPreConfigureEnv,
  // &CustomPostConfigureEnv); RTNAME(RegisterConfigureEnv)(nullptr,
  // &CustomPostConfigureEnv);
  RTNAME(RegisterConfigureEnv)(&CustomPreConfigureEnv, nullptr);
}
} // namespace Fortran::runtime
```

Commit:
[flang-rt] Add the ability to have user supplied callback functions to further customize the runtime environment.

	modified:   flang-rt/include/flang-rt/runtime/environment.h
	modified:   flang-rt/lib/runtime/environment.cpp

vzakhari commented on flang-rt/include/flang-rt/runtime/environment.h:
```diff
@@ -42,6 +55,14 @@ struct ExecutionEnvironment {
+
+  // Optional callback routines to be invoked pre and post
+  // execution environment setup.
+  void (*PreConfigureEnv)(int, const char *[], const char *[],
```
Thank you, Dave!

It looks good to me, except maybe we should typedef the type of the callback functions (probably within `ExecutionEnvironment` struct), so that we can reduce text here and in the declaration of `RegisterConfigureEnv`.

Commit:
[flang-rt] Incorporate review comment #155646.
	modified:   ../../include/flang-rt/runtime/environment.h
	modified:   environment.cpp

Commit:
[flang-rt] Cleanup dead code.
	modified:   ../../include/flang-rt/runtime/environment.h

Commit:
[flang-rt] Run post environment configuration callback functions in the order they are registered.
	modified:   environment.cpp

klausler commented on flang-rt/include/flang-rt/runtime/environment.h:
```diff
@@ -35,13 +36,33 @@ RT_API_ATTRS Fortran::common::optional<Convert> GetConvertFromString(
+  // for either pre or post functions.
+
+  static constexpr int nConfigEnvCallback{8};
+  int nPreConfigEnvCallback{0};
```
I don't think that you want to define a variable like this in a header file.

klausler commented on flang-rt/lib/runtime/environment.cpp:
```diff
@@ -78,6 +78,15 @@ void ExecutionEnvironment::Configure(int ac, const char *av[],
+  if (0 != nPreConfigEnvCallback) {
+    // Run an optional callback function after the core of the
+    // ExecutionEnvironment() logic.
+    for (auto i = 0; i != nPreConfigEnvCallback; ++i) {
```
`int i{0};` please in the runtime

d-parks commented on flang-rt/lib/runtime/environment.cpp:
```diff
@@ -249,4 +262,22 @@ std::int32_t ExecutionEnvironment::UnsetEnv(
+// The pre and post callback functions are called upon entry and exit
+// of ExecutionEnvironment::Configure() respectively.
+
+void RTNAME(RegisterConfigureEnv)(
```
In the example above, the constructor will be called before `_QQmain`.

klausler replied:
Yes.  Where from?

Commit:
[flang-rt] Incorporate latest review comments.
	modified:   ../../include/flang-rt/runtime/environment.h
	modified:   environment.cpp

Commit:
[flang-rt] Correct typo.
	modified:   environment.cpp
